src/core/common/as_core_type.hpp

```diff
@@ -0,0 +1,256 @@
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file includes helper methods to cast between public OT C structs and corresponding core C++ classes.
+ */
+
+#ifndef AS_CORE_TYPE_HPP_
+#define AS_CORE_TYPE_HPP_
+
+#include "openthread-core-config.h"
+
+#include "common/instance.hpp"
+#include "crypto/sha256.hpp"
+#include "mac/mac_filter.hpp"
+#include "net/dns_types.hpp"
+
+namespace ot {
+
+/**
+ * @def AsCoreType
+ *
+ * Convert a pointer to public OT type (C struct) to the corresponding code C++ type reference.
+ *
+ * There are two flavors of this method, one dealing with `const` and the other with non-`const` pointers/references.
+ *
+ * @param[in] aObject   A pointer to the object to convert.
+ *
+ * @returns A reference of the corresponding C++ type matching @p aObject.
+ *
+ */
+
+/**
+ * @def AsCoreTypePtr
+ *
+ * Convert a pointer to public OT type (C struct) to the corresponding code C++ type pointer.
+ *
+ * There are two flavors of this method, one dealing with `const` and the other with non-`const` pointers.
+ *
+ * @param[in] aObject   A pointer to the object to convert.
+ *
+ * @returns A pointer of the corresponding C++ type matching @p aObject.
+ *
+ */
+
+/**
+ * @def MapEnum
+ *
+ * Convert between two enumeration types (one public OT type and one core type).
+ *
+ * @param[in] aValue   The enumeration value to convert
+ *
+ * @returns The matching enumeration value.
+ *
+ */
+
+// Helper macro to add `AsCoreType()` and `AsCoreTypePtr` mapping `Type` to `CoreType`.
+#define DefineAsCoreType(Type, CoreType)                                                                         \
+    inline CoreType &      AsCoreType(Type *aObject) { return *static_cast<CoreType *>(aObject); }               \
+    inline CoreType *      AsCoreTypePtr(Type *aObject) { return static_cast<CoreType *>(aObject); }             \
+    inline const CoreType &AsCoreType(const Type *aObject) { return *static_cast<const CoreType *>(aObject); }   \
+    inline const CoreType *AsCoreTypePtr(const Type *aObject) { return static_cast<const CoreType *>(aObject); } \
+    static_assert(true, "") /* To consume `;` after the macro */
```

Comment:
Is it better to use a more c++ style way? I think it can be more extensible.

```suggestion
  template <typename T> struct PublicTypeTrait;

  #define DefineAsCoreType(aType, aCoreType)    \
      template <> struct PublicTypeTrait<aType> \
      {                                         \
          using CoreType = aCoreType;           \
      }
```

Add the end of the file, define template functions:

```cpp
  template <typename Type> inline typename PublicTypeTrait<Type>::CoreType &AsCoreType(Type *aObject)
  {
      return *static_cast<typename PublicTypeTrait<Type>::CoreType *>(aObject);
  }
  template <typename Type> inline typename PublicTypeTrait<Type>::CoreType *AsCoreTypePtr(Type *aObject)
  {
      return static_cast<typename PublicTypeTrait<Type>::CoreType *>(aObject);
  }

  template <typename Type> inline const typename PublicTypeTrait<Type>::CoreType &AsCoreType(const Type *aObject)
  {
      return *static_cast<const typename PublicTypeTrait<Type>::CoreType *>(aObject);
  }
  template <typename Type> inline const typename PublicTypeTrait<Type>::CoreType *AsCoreTypePtr(const Type *aObject)
  {
      return static_cast<const typename PublicTypeTrait<Type>::CoreType *>(aObject);
  }

```

Reply:
Good suggestion. Thanks @bukepo. Changed this in new push.